### PR TITLE
Move object-storage specific code from `IcebergConfigurer` into `ObjectIO` implementations

### PR DESCRIPTION
**Pull request by snazy**

### CODE REVIEW

File: catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergConfigurer.java [hunk 1]
```diff
@@ -339,17 +245,77 @@ IcebergTableConfig icebergConfigPerTable(
...
+
+          String normalizedWarehouseLocation = normalizeS3Scheme(warehouseLocation);
+
+          // Must use both 's3.signer.uri' and 's3.signer.endpoint', because Iceberg before 1.5.0
+          // does
+          // not handle full URIs passed via 's3.signer.endpoint'. This was changed via
+          // https://github.com/apache/iceberg/pull/8976/files#diff-1f7498b6989fffc169f7791292ed2ccb35b305f6a547fd832f6724057c8aca8bR213-R216,
+          // first released in Iceberg 1.5.0. It's unclear how other language implementations deal
+          // with this.
+          config.put(S3_SIGNER_URI, uriInfo.icebergBaseURI().toString());
```
Comment (dimas-b) [inline catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergConfigurer.java:280]: Hmmm, changing S3 `config` inside a "predicate" looks pretty obscure to me... Especially considering that some `ObjectIO` implementations do not call it at all.

I'd also move the setting of `S3_REMOTE_SIGNING_ENABLED` here.

Also, the predicate checks `locations`, which we already know... why add a round trip through a `Predicate`?

Comment (dimas-b) [reply catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergConfigurer.java:280]: Maybe add a separate `ObjectIO.configureRemoteSigning()` method?

File: catalog/files/api/src/main/java/org/projectnessie/catalog/files/api/ObjectIO.java [hunk 2]
```diff
@@ -29,4 +32,22 @@ public interface ObjectIO {
   OutputStream writeObject(StorageUri uri) throws IOException;
 
   void deleteObjects(List<StorageUri> uris) throws IOException;
+
+  void icebergWarehouseConfig(
+      StorageUri warehouse,
+      BiConsumer<String, String> defaultConfig,
+      BiConsumer<String, String> configOverride);
+
+  void icebergTableConfig(
```
Comment (dimas-b) [inline catalog/files/api/src/main/java/org/projectnessie/catalog/files/api/ObjectIO.java:41]: `ObjectIO` is kind of Nessie's internal access channel to storage... Why overload it with Iceberg FileIO stuff? How about adding a separate interface (e.g. `IcebergIOConfigurer`) for the new methods?

File: catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergConfigurer.java (hunk 1 above)
Comment (snazy) [reply catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergConfigurer.java:280]: All the URI stuff - that definitely doesn't belong to ObjectIO. Too complex to get it into ObjectIO

File: catalog/files/impl/src/main/java/org/projectnessie/catalog/files/adls/AdlsObjectIO.java [hunk 3]
```diff
@@ -101,4 +113,100 @@ public void deleteObjects(List<StorageUri> uris) throws IOException {
...
+
+    if (forTable) {
+      clientSupplier
+          .generateUserDelegationSas(storageLocations, fileSystemOptions)
+          .ifPresent(sasToken -> config.accept(ADLS_SAS_TOKEN_PREFIX + storageAccount, sasToken));
+    }
+  }
+
+  void icebergConfigDefaults(BiConsumer<String, String> config) {
+    config.accept(FILE_IO_IMPL, "org.apache.iceberg.azure.adlsv2.ADLSFileIO");
```
[the remark this answers is not in the document]
Comment (snazy) [reply catalog/files/impl/src/main/java/org/projectnessie/catalog/files/adls/AdlsObjectIO.java:210]: We usually expose vended credentials, which are object storage dependent anyway.
BTW: That's not new to this change.

File: catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergConfigurer.java (hunk 1 above)
Comment (dimas-b) [reply catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergConfigurer.java:280]: I'm +1 to keep it out of `ObjectIO`. 

Still, WDYT about removing the `Predicate` parameter and instead doing `if(isS3(locations)) {...}` here (including the setting of `S3_REMOTE_SIGNING_ENABLED`)?

Comment (snazy) [reply catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergConfigurer.java]: Not really possible - see the comment I've added

**Review by dimas-b — APPROVED** [review]

**Pull request merged**